File: core/src/main/java/com/pluxity/asset/dto/AssetResponse.java
package com.pluxity.asset.dto;

import com.fasterxml.jackson.annotation.JsonUnwrapped;
import com.pluxity.asset.entity.Asset;
import com.pluxity.file.dto.FileResponse;
import com.pluxity.global.response.BaseResponse;

public record AssetResponse(
        Long id,
        String name,
        String code,
        Long categoryId,
        String categoryName,
        String categoryCode,
        FileResponse file,
        FileResponse thumbnailFile,
        @JsonUnwrapped BaseResponse baseResponse) {

    public static AssetResponse from(Asset asset, FileResponse file, FileResponse thumbnailFile) {
        return new AssetResponse(
                asset.getId(),
                asset.getName(),
                asset.getCode(),
                asset.getCategory() != null ? asset.getCategory().getId() : null,
                asset.getCategory() != null ? asset.getCategory().getName() : null,
                asset.getCategory() != null ? asset.getCategory().getCode() : null,
                file != null ? file : FileResponse.empty(),
                thumbnailFile != null ? thumbnailFile : FileResponse.empty(),
                BaseResponse.of(asset));
    }

    public static AssetResponse from(Asset asset) {
        return from(asset, null, null);
    }
}

File: common/src/main/java/com/pluxity/asset/dto/AssetResponse.kt
package com.pluxity.asset.dto

import com.fasterxml.jackson.annotation.JsonUnwrapped
import com.pluxity.asset.entity.Asset
import com.pluxity.file.dto.FileResponse
import com.pluxity.global.response.BaseResponse

data class AssetResponse(
    val id: Long?,
    val name: String?,
    val code: String?,
    val categoryId: Long?,
    val categoryName: String?,
    val categoryCode: String?,
    val file: FileResponse?,
    val thumbnailFile: FileResponse?,
    @field:JsonUnwrapped val baseResponse: BaseResponse,
) {
    companion object {
        @JvmStatic
        fun from(
            asset: Asset,
            file: FileResponse?,
            thumbnailFile: FileResponse?,
        ): AssetResponse =
            AssetResponse(
                id = asset.id,
                name = asset.name,
                code = asset.code,
                categoryId = asset.category?.id,
                categoryName = asset.category?.name,
                categoryCode = asset.category?.code,
                file = file ?: FileResponse(),
                thumbnailFile = thumbnailFile ?: FileResponse(),
                baseResponse = BaseResponse.of(asset),
            )

        @JvmStatic
        fun from(asset: Asset): AssetResponse = from(asset, null, null)
    }
}
